File: tsinterop/src/org/jetbrains/dukat/tsinterop/ExportContent.kt
package org.jetbrains.dukat.tsinterop

import java.util.*


open class ExportContent<T>(
        private val exportTable: MutableMap<T, String> = mutableMapOf()
) {
    private fun createKey(node: T): T {
        return node
    }

    fun getUID(node: T): String? {
        val nodeKey = createKey(node)
        return exportTable.getOrPut(nodeKey) { UUID.randomUUID().toString() }
    }
}

File: typescript/converter/src/ExportContent.ts
import {uid} from "./uid";

export class ExportContent<T> {
  constructor(private exportTable: Map<T, string> = new Map()){}

  private createKey(node: T): T {
    return node;
  }

  getUID(node: T): string {
    let nodeKey = this.createKey(node);
    if (!this.exportTable.has(node)) {
      this.exportTable.set(node, uid());
    }
    return this.exportTable.get(node) || "";
  }
}

export function createExportContent(): ExportContent<any> {
  return new ExportContent<any>();
}
